src/icrar/leap-accelerate/tests/math/vector_tests.cc

```diff
@@ -58,119 +83,102 @@ class vector_tests : public testing::Test
-            auto d_a = icrar::cuda::device_vector<int>(a);
-            auto d_b = icrar::cuda::device_vector<int>(b);
-            auto d_c = icrar::cuda::device_vector<int>(c);
-            icrar::cuda::add(d_a, d_b, d_c);
-            d_c.ToHost(c);
-        }
-        else
-        {
-            icrar::cpu::add(a, b, c);
-        }
+        auto d_a = icrar::cuda::device_vector<int>(a);
+        auto d_b = icrar::cuda::device_vector<int>(b);
+        auto d_c = icrar::cuda::device_vector<int>(c);
+        icrar::cuda::add(d_a, d_b, d_c);
+        d_c.ToHost(c);
 
         std::vector<int> expected = std::vector<int>(n, 16);
         ASSERT_EQ(c, expected);
     }
+#endif // USE_CUDA
 
-    void test_device_vector_fibonacci(const int n, const int k, bool useGpu)
+    template<typename backend, typename Container>
+    void _fibonacci(const Container &a, const Container &b, Container &c, int k)
     {
-        std::vector<int> a = std::vector<int>(n, 1);
-        std::vector<int> b = std::vector<int>(n, 1);
-        std::vector<int> out = std::vector<int>(n, 0);
-
-        if(useGpu)
+        auto n1 = &a;
+        auto n2 = &b;
+        auto n3 = &c;
+        for(int i = 0; i < k; i++)
         {
-            auto d_a = icrar::cuda::device_vector<int>(a);
-            auto d_b = icrar::cuda::device_vector<int>(b);
-            auto d_c = icrar::cuda::device_vector<int>(out);
-
-            auto n1 = &d_a;
-            auto n2 = &d_b;
-            auto n3 = &d_c;
-            for(int i = 0; i < k; i++)
-            {
-                icrar::cuda::add(*n1, *n2, *n3);
-                n1 = n2;
-                n2 = n3;
-            }
-            n3->ToHost(out);
-        }
-        else
-        {
-            auto n1 = &a;
-            auto n2 = &b;
-            auto n3 = &out;
-            for(int i = 0; i < k; i++)
-            {
-                icrar::cpu::add(*n1, *n2, *n3);
-                n1 = n2;
-                n2 = n3;
-            }
-            out = *n3;
+            add(backend{}, *n1, *n2, *n3);
+            n1 = n2;
+            n2 = n3;
         }
+    }
 
+    void fibonacci(cpu, const std::vector<int> &a, const std::vector<int> &b, std::vector<int> &out, int k)
+    {
+        _fibonacci<cpu>(a, b, out, k);
+    }
+
+#ifdef USE_CUDA
+    void fibonacci(cuda, const std::vector<int> &a, const std::vector<int> &b, std::vector<int> &out, int k)
+    {
+        auto d_a = icrar::cuda::device_vector<int>(a);
+        auto d_b = icrar::cuda::device_vector<int>(b);
+        auto d_c = icrar::cuda::device_vector<int>(out);
+        _fibonacci<cuda>(d_a, d_b, d_c, k);
+        d_c.ToHost(out);
+    }
+#endif // USE_CUDA
+
+    template<typename backend>
+    void test_device_vector_fibonacci(const int n, const int k)
+    {
+        std::vector<int> a = std::vector<int>(n, 1);
+        std::vector<int> b = std::vector<int>(n, 1);
+        std::vector<int> out = std::vector<int>(n, 0);
+        fibonacci(backend{}, a, b, out, k);
```

Comment:
It's a little odd using empty structs in order to overload a function (fibonacci in this case just an example function for a sequence of operations with gpu memory and how to test it)

If there is no compilation overhead to it I suppose its a valid workaround to using a template argument as a namespace, I'll consider it some more before using it in production functions

Reply:
Yes, it's not great, but given that partial function template specialization doesn't exist this is another way of achieving the same that seems to be popular. There's a way to hide from users those empty structs, but would require an extra layer of code, and since this is test code it doesn't make sense to complicate it even further.

Reply:
@calgray I'm not sure what the outcome of this is -- are you happy to merge this, or you'd rather see some further changes?